### ISSUE TITLE
Chorded snippets (snippets activated by pressing keys simultaneously)

### ISSUE THREAD

**Issue by dtinth** [microsoft/vscode#46487]
Issue Type: <b>Feature Request</b>

I used “chorded snippets” a lot in Vim and Atom. I tried to create an extension in VS Code but since VS Code does not provide extension point for such extension to be possible. This is the missing feature that’s preventing me from switching my main editor to VS Code.

**What are chorded snippets?**

- “Chorded snippets” are snippets activated by pressing multiple keys at the same time, similar to a musical chord<sup>1</sup>.

- For example, I defined a chorded snippet `con` &rarr; `console.log($0)`. I can activate this snippet by pressing the <kbd>c</kbd><kbd>n</kbd><kbd>o</kbd> keys at the same time.

- It also allows for typing text more quickly, e.g. with a chorded snippet `the` &rarr; `the` I can type the word “the” by slamming <kbd>e</kbd><kbd>h</kbd><kbd>t</kbd> key at the same time.

- Here’s a [video demo in Atom](https://www.youtube.com/watch?v=Bn0Z0vqBTmQ).

Now I am trying to fork VS Code and implementing it by myself.

**In other editors...**

- Atom: [dtinth/atom-arpeggio](https://github.com/dtinth/atom-arpeggio)
- Vim: [kana/vim-arpeggio](https://github.com/kana/vim-arpeggio)

**My question:**

Actually I would love to have this feature available, or made possible as an extension.

- Have I missed any API extensibility point? I need access to the individual `keydown` and `keyup` events to implement it as an extension.

- Would you willing to accept a PR that does one of these?

    - Chorded snippets implemented directly inside VS Code, e.g. inside `src/editor/vs/contrib/snippet`.
    - `keydown`/`keyup` events exposed to extensions.
    - `chord` events exposed to extensions. Here’s an example of [chord detection algorithm](https://codepen.io/dtinth/pen/mxMNOj)
    - Anything else I can do to make this possible?

Thank you for reading!

**Footnotes:**

- <sup>1</sup> Now, I understand that this is different from Visual Studio Code’s definition of “chords” which refers to multiple separate keypress actions (e.g. `ctrl+k ctrl+c`), musically known as “broken chords:”

    > A broken chord is a chord broken into a sequence of notes. A broken chord may repeat some of the notes from the chord…

VS Code version: Code 1.21.1 (79b44aa704ce542d8ca4a3cc44cfca566e7720f1, 2018-03-14T14:39:09.335Z)
OS version: Darwin x64 17.3.0

**Comment by alexandrudima** [comment]
You could write an extension which registers (overwrites) the `type` command and then invokes `default:type`. You cannot write logic to express having `c`, `n`, `o` all pressed (i.e. keydown but not keyup yet), but you could approximate your desired feature by looking for the produced text and perhaps the delta time between the invocations.

**Comment by dtinth** [comment]
@alexandrudima Thank you for the comment. Your suggestion is how `vim-arpeggio` is actually implemented. Based on my experience of using it for a few years, it works, but is prone to both false positives and false negatives.

- Typing too fast causes the chorded snippet to be activated unintentionally.

- Chords sometimes fail due to event processing causing the keypress events to be delayed and not recognized as chords.

Then Atom came along and so I wrote my own plugin in Atom and came up with the chord detection algorithm that is more robust. That is why I had to access individual key event to properly implement the extension.
